packages/telemetry/src/flight-recorder.js

```
@@ -0,0 +1,249 @@
+  );
+  assert.equal(
+    arenaSize,
+    hdrArenaSize,
+    X`${bufferFile} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`,
+  );
+  const arena = new Uint8Array(
+    fileBuf.buffer,
+    header.byteLength,
+    Number(arenaSize),
+  );
+
+  /**
+   * @param {Uint8Array} data
+   * @param {number} [offset]
+   * @returns {IteratorResult<Uint8Array, void>}
+   */
+  const readCircBuf = (data, offset = 0) => {
+    assert(
+      offset + data.byteLength <= arenaSize,
+      X`Reading past end of circular buffer`,
+    );
+
+    // Read the data to the end of the arena.
+    let firstReadLength = data.byteLength;
+    const circStart = header.getBigUint64(I_CIRC_START);
+    const circEnd = header.getBigUint64(I_CIRC_END);
+    const readStart = (Number(circStart) + offset) % Number(arenaSize);
+    if (circStart > circEnd) {
+      // The data is wrapped around the end of the arena, like BBB---AAA
+      firstReadLength = Math.min(
+        firstReadLength,
+        Number(arenaSize) - readStart,
+      );
+      if (readStart >= circEnd && readStart < circStart) {
+        return { done: true, value: undefined };
+      }
+    } else if (readStart < circStart || readStart >= circEnd) {
+      // The data is contiguous, like ---AAABBB---
+      return { done: true, value: undefined };
+    }
+    data.set(arena.subarray(readStart, readStart + firstReadLength));
+    if (firstReadLength < data.byteLength) {
+      data.set(
+        arena.subarray(0, data.byteLength - firstReadLength),
+        firstReadLength,
+      );
+    }
+    return { done: false, value: data };
+  };
+
+  /** @param {Uint8Array} data */
+  const writeCircBuf = data => {
+    if (BigUint64Array.BYTES_PER_ELEMENT + data.byteLength > arena.byteLength) {
+      // The data is too big to fit in the arena, so skip it.
+      const tooBigRecord = JSON.stringify({
+        type: 'slog-record-too-big',
+        size: data.byteLength,
```

I'm tempted to include the first N bytes of the thing that was too big, but that's no guarantee that we capture enough information to figure out what the problem was. But, our slogged objects usually have `type:` at the beginning, so the chances are good.

I'd repeat the size check again after this replacement: if even the `too-big` message is too big, I'd rather silently drop the record than corrupt the file. This would only happen if you configured a ridiculously small file, but this code doesn't know that.

Are you embedding/relying-upon the fact that the per-record header is exactly one 64-bit word long? Maybe factor that out into a `HEADER_SIZE` so it's obvious when reading just `writeCircBuf`.